stan/math/prim/mat/prob/ordered_logistic_glm_lpmf.hpp

```diff
@@ -0,0 +1,174 @@
+#include <stan/math/prim/scal/meta/operands_and_partials.hpp>
+#include <stan/math/prim/scal/err/check_consistent_sizes.hpp>
+#include <stan/math/prim/scal/err/check_bounded.hpp>
+#include <stan/math/prim/scal/err/check_finite.hpp>
+#include <stan/math/prim/mat/fun/value_of_rec.hpp>
+#include <stan/math/prim/arr/fun/value_of_rec.hpp>
+#include <stan/math/prim/mat/meta/as_column_vector_or_scalar.hpp>
+#include <stan/math/prim/scal/meta/as_column_vector_or_scalar.hpp>
+#include <stan/math/prim/scal/meta/include_summand.hpp>
+#include <stan/math/prim/scal/fun/size_zero.hpp>
+#include <stan/math/prim/mat/err/check_ordered.hpp>
+#include <stan/math/prim/arr/err/check_ordered.hpp>
+#include <stan/math/prim/mat/fun/log1m_exp.hpp>
+#include <cmath>
+
+namespace stan {
+namespace math {
+
+/**
+ * Returns the log PMF of the ordinal regression Generalized Linear Model (GLM).
+ * This is equivalent to and faster than ordered_logistic_lpmf(y, x * beta,
+ * cuts).
+ *
+ * @tparam T_y type of integer vector of dependent variables (labels)
+ * @tparam T_x_scalar type of elements in the matrix of independent variables
+ * (features)
+ * @tparam T_beta T_beta type of the weight vector
+ * @tparam T_cuts T_beta type of the cutpoints vector
+ * @param y integer vector parameter
+ * @param x design matrix
+ * @param beta weight vector
+ * @param cuts vector of cutpoints
+ * @return log probability
+ * @throw std::domain_error If the outcome is not between 1 and
+ * the number of cutpoints plus 2; or if the cutpoint vector is not sorted in
+ * ascending order; or any input is not finite
+ * @throw std::invalid_argument If y and lambda are different
+ * lengths.
+ * @throw std::invalid_argument if container sizes mismatch.
+ */
+template <bool propto, typename T_y, typename T_x_scalar, typename T_beta,
+          typename T_cuts>
+typename stan::return_type<T_x_scalar, T_beta, T_cuts>::type
+ordered_logistic_glm_lpmf(
+    const T_y& y,
+    const Eigen::Matrix<T_x_scalar, Eigen::Dynamic, Eigen::Dynamic>& x,
+    const T_beta& beta, const T_cuts& cuts) {
+  static const char* function = "ordered_logistic_glm_lpmf";
+  typedef typename partials_return_type<T_y, T_x_scalar, T_beta, T_cuts>::type
+      T_partials_return;
+
+  using Eigen::Array;
+  using Eigen::Dynamic;
+  using Eigen::Matrix;
+  using Eigen::VectorXd;
+  using std::exp;
+
+  const size_t N_instances = x.rows();
+  const size_t N_attributes = x.cols();
+  const size_t N_classes = length(cuts) + 1;
+
+  check_consistent_size(function, "Vector of dependent variables", y,
+                        N_instances);
+  check_consistent_size(function, "Weight vector", beta, N_attributes);
+  check_bounded(function, "Vector of dependent variables", y, 1, N_classes);
+  check_ordered(function, "Cut-points", cuts);
+  check_finite(function, "Final cut-point", cuts[N_classes - 2]);
+  check_finite(function, "First cut-point", cuts[0]);
+
+  if (size_zero(y, x, beta, cuts))
```

Comment:
This isn't your fault, but it's another function that landed without sufficient review---it should be named `size_zero_any` or something like that.  Anyway, guess that's a different PR.

Reply:
Yeah, this one is on me. This got in before the _any discussion in the `is_nan` PR. Sorry about that. I am assigning myself to fix this one.